File: C++/e4.cpp
#include <iostream>

using namespace std;

// determines if 6 digit number is palindrome
bool palindrome(int x)
{
    int digits[6];
    for(int i = 0; i < 6; i++)
    {
        digits[i] = x % 10;
        x = (x - digits[i]) / 10;
    }
    if(digits[0] == digits[5] && digits[1] == digits[4] && digits[2] == digits[3])
        return true;
    return false;
}

int main()
{
    int largest = 0;
    for(int i = 1; i < 1000; i++)
    {
        for(int j = i; j < 1000; j++)
        {
            if(largest < i * j && palindrome(i * j))
            {
                largest = i * j;
            }
        }
    }
    cout << largest << endl;
}

File: Haskell/e4.hs
-- makes a list of the digits of a number (reversed)

digits :: (Integral a) => a -> [a]
digits 0 = []
digits n =
    let r = n `mod` 10 -- last digit
    in r : (digits $ (n - r) `div` 10) -- recursively apply digits to the first digits

-- evaluates if x is a palindrome
isPalindrome :: (Integral a) => a -> Bool
isPalindrome x = (digits x) == (reverse $ digits x)

-- list of palindromes that are a product of two 3 digit numbels
palindromes :: [Int]
palindromes = filter isPalindrome [x * y | x <- [100..999], y <- [100..999]]

-- take maximum
result :: Int
result = maximum palindromes
